support/elasticsearch2.erl

```diff
@@ -196,3 +265,101 @@ delete_doc(DocId, Context) ->
+    Commands :: [ bulkcmd() ],
+    Context :: z:context(),
+    Result :: result().
+bulk(Commands, Context) ->
+    Connection = connection(Context),
+    elasticsearch2_fetch:bulk(Connection, Commands).
+
+
+%% @doc Using the scrolling API, loop over all documents in an index.
+%% The Accumulator is passed to the function, and the end-accumulator
+%% is returned.
+-spec fold(Index, Fun, Accumulator, Context) -> Result when
+    Index :: index(),
+    Fun :: fun( (Doc, Accumulator1, Context) -> any() )
+         | fun( (Doc, Context) -> any() ),
+    Doc :: map(),
+    Context :: any(),
+    Result :: {ok, Accumulator2} | {error, term()},
+    Accumulator :: any(),
+    Accumulator1 :: any(),
+    Accumulator2 :: any().
+fold(Index, Fun, Acc, Context) ->
+    Query = #{
+        <<"size">> => 100,
+        <<"sort">> => [
+            <<"_doc">>
+        ]
+    },
+    fold(Index, Query, Fun, Acc, Context).
+
+-spec fold(Index, Query, Fun, Accumulator, Context) -> Result when
+    Index :: index(),
+    Query :: map(),
+    Fun :: fun( (Doc, Accumulator1, Context) -> any() )
+         | fun( (Doc, Context) -> any() ),
+    Doc :: map(),
+    Context :: any(),
+    Result :: {ok, Accumulator2} | {error, term()},
+    Accumulator :: any(),
+    Accumulator1 :: any(),
+    Accumulator2 :: any().
+fold(Index, Query, Fun, Acc, Context) ->
+    Result = search_scroll(Index, Query, <<"1h">>, Context),
+    fold_loop(Result, Fun, Acc, Context).
+
+fold_loop({ok, #{
+        <<"hits">> := #{
+            <<"hits">> := []
+        }
+    }}, _Fun, Acc, _Context) ->
+    {ok, Acc};
+fold_loop({ok, #{
+        <<"_scroll_id">> := ScrollId,
+        <<"hits">> := #{
+            <<"hits">> := Docs,
+            <<"total">> := _Total
+        }
+    }}, Fun, Acc, Context) ->
+    % Handle Docs
+    lager:debug("fetched ~p docs", [ length(Docs) ]),
+    Acc1 = lists:foldl(
+        fun(Doc, DocAcc) ->
+            if
+                is_function(Fun, 2) ->
+                    Fun(Doc, Context),
+                    DocAcc;
+                is_function(Fun, 3) ->
+                    Fun(Doc, DocAcc, Context)
+            end
+        end,
+        Acc,
+        Docs),
+    Query = #{
+        <<"scroll">> => <<"10m">>,
+        <<"scroll_id">> => ScrollId
+    },
+    Next = search_scroll_next(Query, Context),
+    fold_loop(Next, Fun, Acc1, Context);
+fold_loop({ok, Other}, _Fun, _Acc, _Context) ->
+    lager:warning("Scroll returned unexpected: ~p", [ Other ]),
+    {error, Other};
+fold_loop({error, Reason} = Error, _Fun, _Acc, _Context) ->
+    lager:error("Scroll returned error: ~p", [ Reason ]),
+    Error.
+
+-spec search_scroll(binary(), map(), string()|binary(), z:context()) -> {ok, map()} | {error, term()}.
+search_scroll(Index, Query, Timeout, Context) ->
+    lager:debug("mod_elasticsearch2 (~s): ~s", [ Index, Query ]),
+    % io:format("~p:~p: (~s) ~n~p~n~n", [ ?MODULE, ?LINE, Index, ElasticQuery2 ]),
```

Comment:
Gemiste debug log?

Reply:
Replaced with a proper queue.

Reply:
Ik bedoelde de comment, maar okay is niet erg.